Handle Non-Patient Compartment Resource IDs

trobanga opened the pull request:
## Summary

IDs of resources NOT in the patient compartment are now pseudonymized directly via gPAS, instead of being derived from patient ID/salt. This addresses issue #942.

### Changes
- **PatientCompartment utility**: Loads the FHIR R4B patient compartment definition and checks if a resource type has a `param` key (indicating it's patient-related)
- **GpasClient**: Added batch `fetchOrCreatePseudonyms()` method for efficient pseudonymization of multiple IDs
- **FhirMappingProvider**: Splits resource IDs by compartment membership:
  - Patient-compartment resources (66 types): Continue using salt-based SHA256 hashing
  - Non-compartment resources (Organization, Practitioner, Medication, etc.): Direct gPAS pseudonymization

### Technical Details
- Resource ID format: `{patientId}.{ResourceType}:{id}`
- Identifier format: `{patientId}.identifier.{system}:{value}` (always treated as patient-related)
- Uses embedded `compartmentdefinition-patient.json` from FHIR R4B spec

## Test Plan
- [x] All existing tests pass (85 tests in trust-center-agent)
- [x] New unit tests for PatientCompartment (6 tests)
- [x] New unit tests for extractResourceType and generateNonCompartmentMapping (7 tests)
- [x] Integration test with actual gPAS instance

Closes #942

trobanga committed:
Handle Non-Patient Compartment Resource IDs

Resource IDs outside the patient compartment are now pseudonymized
directly via gPAS instead of using patient-derived salt.

- Add PatientCompartment to check FHIR compartment membership
- Add batch fetchOrCreatePseudonyms to GpasClient
- Split ID handling in FhirMappingProvider by compartment type

Closes #942

trobanga committed:
Extract CompartmentIdSplitter from FhirMappingProvider

Reduce FhirMappingProvider from 321 to 189 lines by extracting
compartment-related logic into a dedicated class.

trobanga committed:
Refactor FhirMappingProvider and PatientCompartment

trobanga committed:
Move Compartment Membership Check to CDA

CDA now determines actual compartment membership by checking
if resource param fields reference the patient. TCA uses
prefix-based checking only.

trobanga committed:
Pass Patient Resource ID Through Pipeline

trobanga committed:
Replace MySQL with MariaDB for gPAS DB

- Switch gpas-db from MySQL 9.5 to MariaDB 11.4
- Add custom gPAS Dockerfile with MariaDB JDBC driver
- Update environment variables and healthcheck for MariaDB

trobanga committed:
Revert Timeout and Fix MySQL Image Hash

trobanga committed:
Add Tests for Reference Edge Cases

trobanga committed:
Add IdatScraper Non-Bundle Branch Tests

trobanga committed:
Remove Unreachable Null Check

trobanga committed:
Remove Defensive hasResource() Check

trobanga committed:
Remove Dead Code Path from IdatScraper

trobanga committed:
Add Config Option for Compartment ID Source

trobanga committed:
Rename Config to enableCompartmentNamespacing

trobanga committed:
Make GpasClient Batch Size Configurable

trobanga committed:
Make CompartmentIdSplitter Static Utility

trobanga committed:
Convert CompartmentIdSplitter To Interface

knoppiks commented on clinical-domain-agent/src/main/java/care/smith/fts/cda/services/configuration/PatientCompartmentServiceConfiguration.java:
```diff
@@ -0,0 +1,63 @@
+    var resourceTypeToParams = loadCompartmentDefinition(objectMapper);
+    log.info(
+        "Loaded patient compartment service with {} resource types having compartment params",
+        resourceTypeToParams.values().stream().filter(params -> !params.isEmpty()).count());
```
```suggestion
        resourceTypeToParams.values().stream().filter(not(List::isEmpty)).count());
```

knoppiks commented on clinical-domain-agent/src/main/java/care/smith/fts/cda/services/deidentifhir/CompartmentMembershipChecker.java:
```diff
@@ -0,0 +1,179 @@
+      return false;
+    }
+
+    for (String param : params) {
```
Please use a Stream.filter here.

knoppiks commented on clinical-domain-agent/src/main/java/care/smith/fts/cda/services/deidentifhir/CompartmentMembershipChecker.java:
```diff
@@ -0,0 +1,179 @@
+      return true;
+    }
+
+    // Also check if the reference ends with just the patient ID after "Patient/"
```
Isn't that redundant?

knoppiks commented on clinical-domain-agent/src/main/java/care/smith/fts/cda/services/PatientCompartmentService.java:
```diff
@@ -0,0 +1,41 @@
+ * can link the resource to a patient. For example, ServiceRequest has params ["subject",
+ * "performer"], meaning it's in the patient compartment if either field references the patient.
+ */
+public class PatientCompartmentService {
```
This service does nearly nothing. Please rename to `PatientCompartmentRepository` - move the creation of the `resourceTypeToParams` Map into a static method, incorporate the PatientCompartmentChecker functionality.

trobanga committed:
Revert MariaDB Changes for gPAS DB

trobanga committed:
Remove Redundant GpasClientTest

trobanga committed:
Add Nested Path Resolution For Compartment Params

Expand compartment membership checking to handle nested FHIR paths
(e.g., Appointment.participant.actor, CareTeam.participant.member)
and simple field name mappings (e.g., policy-holder -> policyHolder).

trobanga committed:
Add Tests For Nested Path Edge Cases

trobanga committed:
Rename To PatientCompartmentService

trobanga committed:
Remove getCompartmentDefinitionPath Method

trobanga committed:
Inline PATIENT_ID And Format Code

trobanga committed:
Remove Redundant Test Assertions

trobanga committed:
Refactor To Functional Style With Optional

trobanga committed:
Separate Compartment IDs At Source In CDA
